// @aave/core-v3/types/protocol/libraries/helpers/Errors.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {FunctionFragment, Result} from "@ethersproject/abi";
import type {Listener, Provider} from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "../../../common";

export interface ErrorsInterface extends utils.Interface {
  functions: {
    "ACL_ADMIN_CANNOT_BE_ZERO()": FunctionFragment;
    "ADDRESSES_PROVIDER_ALREADY_ADDED()": FunctionFragment;
    "ADDRESSES_PROVIDER_NOT_REGISTERED()": FunctionFragment;
    "AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE()": FunctionFragment;
    "ASSET_NOT_BORROWABLE_IN_ISOLATION()": FunctionFragment;
    "ASSET_NOT_LISTED()": FunctionFragment;
    "BORROWING_NOT_ENABLED()": FunctionFragment;
    "BORROW_CAP_EXCEEDED()": FunctionFragment;
    "BRIDGE_PROTOCOL_FEE_INVALID()": FunctionFragment;
    "CALLER_MUST_BE_POOL()": FunctionFragment;
    "CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN()": FunctionFragment;
    "CALLER_NOT_ATOKEN()": FunctionFragment;
    "CALLER_NOT_BRIDGE()": FunctionFragment;
    "CALLER_NOT_EMERGENCY_ADMIN()": FunctionFragment;
    "CALLER_NOT_POOL_ADMIN()": FunctionFragment;
    "CALLER_NOT_POOL_CONFIGURATOR()": FunctionFragment;
    "CALLER_NOT_POOL_OR_EMERGENCY_ADMIN()": FunctionFragment;
    "CALLER_NOT_RISK_OR_POOL_ADMIN()": FunctionFragment;
    "COLLATERAL_BALANCE_IS_ZERO()": FunctionFragment;
    "COLLATERAL_CANNOT_BE_LIQUIDATED()": FunctionFragment;
    "COLLATERAL_CANNOT_COVER_NEW_BORROW()": FunctionFragment;
    "COLLATERAL_SAME_AS_BORROWING_CURRENCY()": FunctionFragment;
    "DEBT_CEILING_EXCEEDED()": FunctionFragment;
    "DEBT_CEILING_NOT_ZERO()": FunctionFragment;
    "EMODE_CATEGORY_RESERVED()": FunctionFragment;
    "FLASHLOAN_DISABLED()": FunctionFragment;
    "FLASHLOAN_PREMIUM_INVALID()": FunctionFragment;
    "HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD()": FunctionFragment;
    "HEALTH_FACTOR_NOT_BELOW_THRESHOLD()": FunctionFragment;
    "INCONSISTENT_EMODE_CATEGORY()": FunctionFragment;
    "INCONSISTENT_FLASHLOAN_PARAMS()": FunctionFragment;
    "INCONSISTENT_PARAMS_LENGTH()": FunctionFragment;
    "INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET()": FunctionFragment;
    "INVALID_ADDRESSES_PROVIDER()": FunctionFragment;
    "INVALID_ADDRESSES_PROVIDER_ID()": FunctionFragment;
    "INVALID_AMOUNT()": FunctionFragment;
    "INVALID_BORROW_CAP()": FunctionFragment;
    "INVALID_BURN_AMOUNT()": FunctionFragment;
    "INVALID_DEBT_CEILING()": FunctionFragment;
    "INVALID_DECIMALS()": FunctionFragment;
    "INVALID_EMODE_CATEGORY()": FunctionFragment;
    "INVALID_EMODE_CATEGORY_ASSIGNMENT()": FunctionFragment;
    "INVALID_EMODE_CATEGORY_PARAMS()": FunctionFragment;
    "INVALID_EXPIRATION()": FunctionFragment;
    "INVALID_FLASHLOAN_EXECUTOR_RETURN()": FunctionFragment;
    "INVALID_INTEREST_RATE_MODE_SELECTED()": FunctionFragment;
    "INVALID_LIQUIDATION_PROTOCOL_FEE()": FunctionFragment;
    "INVALID_LIQ_BONUS()": FunctionFragment;
    "INVALID_LIQ_THRESHOLD()": FunctionFragment;
    "INVALID_LTV()": FunctionFragment;
    "INVALID_MINT_AMOUNT()": FunctionFragment;
    "INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO()": FunctionFragment;
    "INVALID_OPTIMAL_USAGE_RATIO()": FunctionFragment;
    "INVALID_RESERVE_FACTOR()": FunctionFragment;
    "INVALID_RESERVE_INDEX()": FunctionFragment;
    "INVALID_RESERVE_PARAMS()": FunctionFragment;
    "INVALID_SIGNATURE()": FunctionFragment;
    "INVALID_SUPPLY_CAP()": FunctionFragment;
    "INVALID_UNBACKED_MINT_CAP()": FunctionFragment;
    "LTV_VALIDATION_FAILED()": FunctionFragment;
    "NOT_CONTRACT()": FunctionFragment;
    "NOT_ENOUGH_AVAILABLE_USER_BALANCE()": FunctionFragment;
    "NO_DEBT_OF_SELECTED_TYPE()": FunctionFragment;
    "NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF()": FunctionFragment;
    "NO_MORE_RESERVES_ALLOWED()": FunctionFragment;
    "NO_OUTSTANDING_STABLE_DEBT()": FunctionFragment;
    "NO_OUTSTANDING_VARIABLE_DEBT()": FunctionFragment;
    "OPERATION_NOT_SUPPORTED()": FunctionFragment;
    "POOL_ADDRESSES_DO_NOT_MATCH()": FunctionFragment;
    "PRICE_ORACLE_SENTINEL_CHECK_FAILED()": FunctionFragment;
    "RESERVE_ALREADY_ADDED()": FunctionFragment;
    "RESERVE_ALREADY_INITIALIZED()": FunctionFragment;
    "RESERVE_DEBT_NOT_ZERO()": FunctionFragment;
    "RESERVE_FROZEN()": FunctionFragment;
    "RESERVE_INACTIVE()": FunctionFragment;
    "RESERVE_LIQUIDITY_NOT_ZERO()": FunctionFragment;
    "RESERVE_PAUSED()": FunctionFragment;
    "SILOED_BORROWING_VIOLATION()": FunctionFragment;
    "SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER()": FunctionFragment;
    "STABLE_BORROWING_ENABLED()": FunctionFragment;
    "STABLE_BORROWING_NOT_ENABLED()": FunctionFragment;
    "STABLE_DEBT_NOT_ZERO()": FunctionFragment;
    "SUPPLY_CAP_EXCEEDED()": FunctionFragment;
    "UNBACKED_MINT_CAP_EXCEEDED()": FunctionFragment;
    "UNDERLYING_BALANCE_ZERO()": FunctionFragment;
    "UNDERLYING_CANNOT_BE_RESCUED()": FunctionFragment;
    "UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO()": FunctionFragment;
    "USER_IN_ISOLATION_MODE()": FunctionFragment;
    "VARIABLE_DEBT_SUPPLY_NOT_ZERO()": FunctionFragment;
    "ZERO_ADDRESS_NOT_VALID()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "ACL_ADMIN_CANNOT_BE_ZERO"
      | "ADDRESSES_PROVIDER_ALREADY_ADDED"
      | "ADDRESSES_PROVIDER_NOT_REGISTERED"
      | "AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE"
      | "ASSET_NOT_BORROWABLE_IN_ISOLATION"
      | "ASSET_NOT_LISTED"
      | "BORROWING_NOT_ENABLED"
      | "BORROW_CAP_EXCEEDED"
      | "BRIDGE_PROTOCOL_FEE_INVALID"
      | "CALLER_MUST_BE_POOL"
      | "CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN"
      | "CALLER_NOT_ATOKEN"
      | "CALLER_NOT_BRIDGE"
      | "CALLER_NOT_EMERGENCY_ADMIN"
      | "CALLER_NOT_POOL_ADMIN"
      | "CALLER_NOT_POOL_CONFIGURATOR"
      | "CALLER_NOT_POOL_OR_EMERGENCY_ADMIN"
      | "CALLER_NOT_RISK_OR_POOL_ADMIN"
      | "COLLATERAL_BALANCE_IS_ZERO"
      | "COLLATERAL_CANNOT_BE_LIQUIDATED"
      | "COLLATERAL_CANNOT_COVER_NEW_BORROW"
      | "COLLATERAL_SAME_AS_BORROWING_CURRENCY"
      | "DEBT_CEILING_EXCEEDED"
      | "DEBT_CEILING_NOT_ZERO"
      | "EMODE_CATEGORY_RESERVED"
      | "FLASHLOAN_DISABLED"
      | "FLASHLOAN_PREMIUM_INVALID"
      | "HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD"
      | "HEALTH_FACTOR_NOT_BELOW_THRESHOLD"
      | "INCONSISTENT_EMODE_CATEGORY"
      | "INCONSISTENT_FLASHLOAN_PARAMS"
      | "INCONSISTENT_PARAMS_LENGTH"
      | "INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET"
      | "INVALID_ADDRESSES_PROVIDER"
      | "INVALID_ADDRESSES_PROVIDER_ID"
      | "INVALID_AMOUNT"
      | "INVALID_BORROW_CAP"
      | "INVALID_BURN_AMOUNT"
      | "INVALID_DEBT_CEILING"
      | "INVALID_DECIMALS"
      | "INVALID_EMODE_CATEGORY"
      | "INVALID_EMODE_CATEGORY_ASSIGNMENT"
      | "INVALID_EMODE_CATEGORY_PARAMS"
      | "INVALID_EXPIRATION"
      | "INVALID_FLASHLOAN_EXECUTOR_RETURN"
      | "INVALID_INTEREST_RATE_MODE_SELECTED"
      | "INVALID_LIQUIDATION_PROTOCOL_FEE"
      | "INVALID_LIQ_BONUS"
      | "INVALID_LIQ_THRESHOLD"
      | "INVALID_LTV"
      | "INVALID_MINT_AMOUNT"
      | "INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO"
      | "INVALID_OPTIMAL_USAGE_RATIO"
      | "INVALID_RESERVE_FACTOR"
      | "INVALID_RESERVE_INDEX"
      | "INVALID_RESERVE_PARAMS"
      | "INVALID_SIGNATURE"
      | "INVALID_SUPPLY_CAP"
      | "INVALID_UNBACKED_MINT_CAP"
      | "LTV_VALIDATION_FAILED"
      | "NOT_CONTRACT"
      | "NOT_ENOUGH_AVAILABLE_USER_BALANCE"
      | "NO_DEBT_OF_SELECTED_TYPE"
      | "NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF"
      | "NO_MORE_RESERVES_ALLOWED"
      | "NO_OUTSTANDING_STABLE_DEBT"
      | "NO_OUTSTANDING_VARIABLE_DEBT"
      | "OPERATION_NOT_SUPPORTED"
      | "POOL_ADDRESSES_DO_NOT_MATCH"
      | "PRICE_ORACLE_SENTINEL_CHECK_FAILED"
      | "RESERVE_ALREADY_ADDED"
      | "RESERVE_ALREADY_INITIALIZED"
      | "RESERVE_DEBT_NOT_ZERO"
      | "RESERVE_FROZEN"
      | "RESERVE_INACTIVE"
      | "RESERVE_LIQUIDITY_NOT_ZERO"
      | "RESERVE_PAUSED"
      | "SILOED_BORROWING_VIOLATION"
      | "SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER"
      | "STABLE_BORROWING_ENABLED"
      | "STABLE_BORROWING_NOT_ENABLED"
      | "STABLE_DEBT_NOT_ZERO"
      | "SUPPLY_CAP_EXCEEDED"
      | "UNBACKED_MINT_CAP_EXCEEDED"
      | "UNDERLYING_BALANCE_ZERO"
      | "UNDERLYING_CANNOT_BE_RESCUED"
      | "UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO"
      | "USER_IN_ISOLATION_MODE"
      | "VARIABLE_DEBT_SUPPLY_NOT_ZERO"
      | "ZERO_ADDRESS_NOT_VALID"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "ACL_ADMIN_CANNOT_BE_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ADDRESSES_PROVIDER_ALREADY_ADDED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ADDRESSES_PROVIDER_NOT_REGISTERED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ASSET_NOT_BORROWABLE_IN_ISOLATION",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ASSET_NOT_LISTED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "BORROWING_NOT_ENABLED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "BORROW_CAP_EXCEEDED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "BRIDGE_PROTOCOL_FEE_INVALID",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_MUST_BE_POOL",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_NOT_ATOKEN",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_NOT_BRIDGE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_NOT_EMERGENCY_ADMIN",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_NOT_POOL_ADMIN",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_NOT_POOL_CONFIGURATOR",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_NOT_POOL_OR_EMERGENCY_ADMIN",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "CALLER_NOT_RISK_OR_POOL_ADMIN",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "COLLATERAL_BALANCE_IS_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "COLLATERAL_CANNOT_BE_LIQUIDATED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "COLLATERAL_CANNOT_COVER_NEW_BORROW",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "COLLATERAL_SAME_AS_BORROWING_CURRENCY",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "DEBT_CEILING_EXCEEDED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "DEBT_CEILING_NOT_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "EMODE_CATEGORY_RESERVED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "FLASHLOAN_DISABLED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "FLASHLOAN_PREMIUM_INVALID",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "HEALTH_FACTOR_NOT_BELOW_THRESHOLD",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INCONSISTENT_EMODE_CATEGORY",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INCONSISTENT_FLASHLOAN_PARAMS",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INCONSISTENT_PARAMS_LENGTH",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_ADDRESSES_PROVIDER",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_ADDRESSES_PROVIDER_ID",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_AMOUNT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_BORROW_CAP",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_BURN_AMOUNT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_DEBT_CEILING",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_DECIMALS",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_EMODE_CATEGORY",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_EMODE_CATEGORY_ASSIGNMENT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_EMODE_CATEGORY_PARAMS",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_EXPIRATION",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_FLASHLOAN_EXECUTOR_RETURN",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_INTEREST_RATE_MODE_SELECTED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_LIQUIDATION_PROTOCOL_FEE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_LIQ_BONUS",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_LIQ_THRESHOLD",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_LTV",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_MINT_AMOUNT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_OPTIMAL_USAGE_RATIO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_RESERVE_FACTOR",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_RESERVE_INDEX",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_RESERVE_PARAMS",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_SIGNATURE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_SUPPLY_CAP",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "INVALID_UNBACKED_MINT_CAP",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "LTV_VALIDATION_FAILED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "NOT_CONTRACT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "NOT_ENOUGH_AVAILABLE_USER_BALANCE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "NO_DEBT_OF_SELECTED_TYPE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "NO_MORE_RESERVES_ALLOWED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "NO_OUTSTANDING_STABLE_DEBT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "NO_OUTSTANDING_VARIABLE_DEBT",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "OPERATION_NOT_SUPPORTED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "POOL_ADDRESSES_DO_NOT_MATCH",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "PRICE_ORACLE_SENTINEL_CHECK_FAILED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "RESERVE_ALREADY_ADDED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "RESERVE_ALREADY_INITIALIZED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "RESERVE_DEBT_NOT_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "RESERVE_FROZEN",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "RESERVE_INACTIVE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "RESERVE_LIQUIDITY_NOT_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "RESERVE_PAUSED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "SILOED_BORROWING_VIOLATION",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "STABLE_BORROWING_ENABLED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "STABLE_BORROWING_NOT_ENABLED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "STABLE_DEBT_NOT_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "SUPPLY_CAP_EXCEEDED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "UNBACKED_MINT_CAP_EXCEEDED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "UNDERLYING_BALANCE_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "UNDERLYING_CANNOT_BE_RESCUED",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "USER_IN_ISOLATION_MODE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "VARIABLE_DEBT_SUPPLY_NOT_ZERO",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ZERO_ADDRESS_NOT_VALID",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "ACL_ADMIN_CANNOT_BE_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "ADDRESSES_PROVIDER_ALREADY_ADDED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "ADDRESSES_PROVIDER_NOT_REGISTERED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "ASSET_NOT_BORROWABLE_IN_ISOLATION",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "ASSET_NOT_LISTED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "BORROWING_NOT_ENABLED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "BORROW_CAP_EXCEEDED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "BRIDGE_PROTOCOL_FEE_INVALID",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_MUST_BE_POOL",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_NOT_ATOKEN",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_NOT_BRIDGE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_NOT_EMERGENCY_ADMIN",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_NOT_POOL_ADMIN",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_NOT_POOL_CONFIGURATOR",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_NOT_POOL_OR_EMERGENCY_ADMIN",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "CALLER_NOT_RISK_OR_POOL_ADMIN",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "COLLATERAL_BALANCE_IS_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "COLLATERAL_CANNOT_BE_LIQUIDATED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "COLLATERAL_CANNOT_COVER_NEW_BORROW",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "COLLATERAL_SAME_AS_BORROWING_CURRENCY",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "DEBT_CEILING_EXCEEDED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "DEBT_CEILING_NOT_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "EMODE_CATEGORY_RESERVED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "FLASHLOAN_DISABLED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "FLASHLOAN_PREMIUM_INVALID",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "HEALTH_FACTOR_NOT_BELOW_THRESHOLD",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INCONSISTENT_EMODE_CATEGORY",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INCONSISTENT_FLASHLOAN_PARAMS",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INCONSISTENT_PARAMS_LENGTH",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_ADDRESSES_PROVIDER",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_ADDRESSES_PROVIDER_ID",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_AMOUNT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_BORROW_CAP",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_BURN_AMOUNT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_DEBT_CEILING",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_DECIMALS",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_EMODE_CATEGORY",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_EMODE_CATEGORY_ASSIGNMENT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_EMODE_CATEGORY_PARAMS",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_EXPIRATION",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_FLASHLOAN_EXECUTOR_RETURN",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_INTEREST_RATE_MODE_SELECTED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_LIQUIDATION_PROTOCOL_FEE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_LIQ_BONUS",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_LIQ_THRESHOLD",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_LTV",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_MINT_AMOUNT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_OPTIMAL_USAGE_RATIO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_RESERVE_FACTOR",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_RESERVE_INDEX",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_RESERVE_PARAMS",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_SIGNATURE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_SUPPLY_CAP",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "INVALID_UNBACKED_MINT_CAP",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "LTV_VALIDATION_FAILED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "NOT_CONTRACT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "NOT_ENOUGH_AVAILABLE_USER_BALANCE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "NO_DEBT_OF_SELECTED_TYPE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "NO_MORE_RESERVES_ALLOWED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "NO_OUTSTANDING_STABLE_DEBT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "NO_OUTSTANDING_VARIABLE_DEBT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "OPERATION_NOT_SUPPORTED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "POOL_ADDRESSES_DO_NOT_MATCH",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "PRICE_ORACLE_SENTINEL_CHECK_FAILED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "RESERVE_ALREADY_ADDED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "RESERVE_ALREADY_INITIALIZED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "RESERVE_DEBT_NOT_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "RESERVE_FROZEN",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "RESERVE_INACTIVE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "RESERVE_LIQUIDITY_NOT_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "RESERVE_PAUSED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "SILOED_BORROWING_VIOLATION",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "STABLE_BORROWING_ENABLED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "STABLE_BORROWING_NOT_ENABLED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "STABLE_DEBT_NOT_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "SUPPLY_CAP_EXCEEDED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "UNBACKED_MINT_CAP_EXCEEDED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "UNDERLYING_BALANCE_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "UNDERLYING_CANNOT_BE_RESCUED",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "USER_IN_ISOLATION_MODE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "VARIABLE_DEBT_SUPPLY_NOT_ZERO",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "ZERO_ADDRESS_NOT_VALID",
    data: BytesLike
  ): Result;

  events: {};
}

export interface Errors extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: ErrorsInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    ACL_ADMIN_CANNOT_BE_ZERO(overrides?: CallOverrides): Promise<[string]>;

    ADDRESSES_PROVIDER_ALREADY_ADDED(
      overrides?: CallOverrides
    ): Promise<[string]>;

    ADDRESSES_PROVIDER_NOT_REGISTERED(
      overrides?: CallOverrides
    ): Promise<[string]>;

    AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE(
      overrides?: CallOverrides
    ): Promise<[string]>;

    ASSET_NOT_BORROWABLE_IN_ISOLATION(
      overrides?: CallOverrides
    ): Promise<[string]>;

    ASSET_NOT_LISTED(overrides?: CallOverrides): Promise<[string]>;

    BORROWING_NOT_ENABLED(overrides?: CallOverrides): Promise<[string]>;

    BORROW_CAP_EXCEEDED(overrides?: CallOverrides): Promise<[string]>;

    BRIDGE_PROTOCOL_FEE_INVALID(overrides?: CallOverrides): Promise<[string]>;

    CALLER_MUST_BE_POOL(overrides?: CallOverrides): Promise<[string]>;

    CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN(
      overrides?: CallOverrides
    ): Promise<[string]>;

    CALLER_NOT_ATOKEN(overrides?: CallOverrides): Promise<[string]>;

    CALLER_NOT_BRIDGE(overrides?: CallOverrides): Promise<[string]>;

    CALLER_NOT_EMERGENCY_ADMIN(overrides?: CallOverrides): Promise<[string]>;

    CALLER_NOT_POOL_ADMIN(overrides?: CallOverrides): Promise<[string]>;

    CALLER_NOT_POOL_CONFIGURATOR(overrides?: CallOverrides): Promise<[string]>;

    CALLER_NOT_POOL_OR_EMERGENCY_ADMIN(
      overrides?: CallOverrides
    ): Promise<[string]>;

    CALLER_NOT_RISK_OR_POOL_ADMIN(overrides?: CallOverrides): Promise<[string]>;

    COLLATERAL_BALANCE_IS_ZERO(overrides?: CallOverrides): Promise<[string]>;

    COLLATERAL_CANNOT_BE_LIQUIDATED(
      overrides?: CallOverrides
    ): Promise<[string]>;

    COLLATERAL_CANNOT_COVER_NEW_BORROW(
      overrides?: CallOverrides
    ): Promise<[string]>;

    COLLATERAL_SAME_AS_BORROWING_CURRENCY(
      overrides?: CallOverrides
    ): Promise<[string]>;

    DEBT_CEILING_EXCEEDED(overrides?: CallOverrides): Promise<[string]>;

    DEBT_CEILING_NOT_ZERO(overrides?: CallOverrides): Promise<[string]>;

    EMODE_CATEGORY_RESERVED(overrides?: CallOverrides): Promise<[string]>;

    FLASHLOAN_DISABLED(overrides?: CallOverrides): Promise<[string]>;

    FLASHLOAN_PREMIUM_INVALID(overrides?: CallOverrides): Promise<[string]>;

    HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<[string]>;

    HEALTH_FACTOR_NOT_BELOW_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<[string]>;

    INCONSISTENT_EMODE_CATEGORY(overrides?: CallOverrides): Promise<[string]>;

    INCONSISTENT_FLASHLOAN_PARAMS(overrides?: CallOverrides): Promise<[string]>;

    INCONSISTENT_PARAMS_LENGTH(overrides?: CallOverrides): Promise<[string]>;

    INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET(
      overrides?: CallOverrides
    ): Promise<[string]>;

    INVALID_ADDRESSES_PROVIDER(overrides?: CallOverrides): Promise<[string]>;

    INVALID_ADDRESSES_PROVIDER_ID(overrides?: CallOverrides): Promise<[string]>;

    INVALID_AMOUNT(overrides?: CallOverrides): Promise<[string]>;

    INVALID_BORROW_CAP(overrides?: CallOverrides): Promise<[string]>;

    INVALID_BURN_AMOUNT(overrides?: CallOverrides): Promise<[string]>;

    INVALID_DEBT_CEILING(overrides?: CallOverrides): Promise<[string]>;

    INVALID_DECIMALS(overrides?: CallOverrides): Promise<[string]>;

    INVALID_EMODE_CATEGORY(overrides?: CallOverrides): Promise<[string]>;

    INVALID_EMODE_CATEGORY_ASSIGNMENT(
      overrides?: CallOverrides
    ): Promise<[string]>;

    INVALID_EMODE_CATEGORY_PARAMS(overrides?: CallOverrides): Promise<[string]>;

    INVALID_EXPIRATION(overrides?: CallOverrides): Promise<[string]>;

    INVALID_FLASHLOAN_EXECUTOR_RETURN(
      overrides?: CallOverrides
    ): Promise<[string]>;

    INVALID_INTEREST_RATE_MODE_SELECTED(
      overrides?: CallOverrides
    ): Promise<[string]>;

    INVALID_LIQUIDATION_PROTOCOL_FEE(
      overrides?: CallOverrides
    ): Promise<[string]>;

    INVALID_LIQ_BONUS(overrides?: CallOverrides): Promise<[string]>;

    INVALID_LIQ_THRESHOLD(overrides?: CallOverrides): Promise<[string]>;

    INVALID_LTV(overrides?: CallOverrides): Promise<[string]>;

    INVALID_MINT_AMOUNT(overrides?: CallOverrides): Promise<[string]>;

    INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO(
      overrides?: CallOverrides
    ): Promise<[string]>;

    INVALID_OPTIMAL_USAGE_RATIO(overrides?: CallOverrides): Promise<[string]>;

    INVALID_RESERVE_FACTOR(overrides?: CallOverrides): Promise<[string]>;

    INVALID_RESERVE_INDEX(overrides?: CallOverrides): Promise<[string]>;

    INVALID_RESERVE_PARAMS(overrides?: CallOverrides): Promise<[string]>;

    INVALID_SIGNATURE(overrides?: CallOverrides): Promise<[string]>;

    INVALID_SUPPLY_CAP(overrides?: CallOverrides): Promise<[string]>;

    INVALID_UNBACKED_MINT_CAP(overrides?: CallOverrides): Promise<[string]>;

    LTV_VALIDATION_FAILED(overrides?: CallOverrides): Promise<[string]>;

    NOT_CONTRACT(overrides?: CallOverrides): Promise<[string]>;

    NOT_ENOUGH_AVAILABLE_USER_BALANCE(
      overrides?: CallOverrides
    ): Promise<[string]>;

    NO_DEBT_OF_SELECTED_TYPE(overrides?: CallOverrides): Promise<[string]>;

    NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF(
      overrides?: CallOverrides
    ): Promise<[string]>;

    NO_MORE_RESERVES_ALLOWED(overrides?: CallOverrides): Promise<[string]>;

    NO_OUTSTANDING_STABLE_DEBT(overrides?: CallOverrides): Promise<[string]>;

    NO_OUTSTANDING_VARIABLE_DEBT(overrides?: CallOverrides): Promise<[string]>;

    OPERATION_NOT_SUPPORTED(overrides?: CallOverrides): Promise<[string]>;

    POOL_ADDRESSES_DO_NOT_MATCH(overrides?: CallOverrides): Promise<[string]>;

    PRICE_ORACLE_SENTINEL_CHECK_FAILED(
      overrides?: CallOverrides
    ): Promise<[string]>;

    RESERVE_ALREADY_ADDED(overrides?: CallOverrides): Promise<[string]>;

    RESERVE_ALREADY_INITIALIZED(overrides?: CallOverrides): Promise<[string]>;

    RESERVE_DEBT_NOT_ZERO(overrides?: CallOverrides): Promise<[string]>;

    RESERVE_FROZEN(overrides?: CallOverrides): Promise<[string]>;

    RESERVE_INACTIVE(overrides?: CallOverrides): Promise<[string]>;

    RESERVE_LIQUIDITY_NOT_ZERO(overrides?: CallOverrides): Promise<[string]>;

    RESERVE_PAUSED(overrides?: CallOverrides): Promise<[string]>;

    SILOED_BORROWING_VIOLATION(overrides?: CallOverrides): Promise<[string]>;

    SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER(
      overrides?: CallOverrides
    ): Promise<[string]>;

    STABLE_BORROWING_ENABLED(overrides?: CallOverrides): Promise<[string]>;

    STABLE_BORROWING_NOT_ENABLED(overrides?: CallOverrides): Promise<[string]>;

    STABLE_DEBT_NOT_ZERO(overrides?: CallOverrides): Promise<[string]>;

    SUPPLY_CAP_EXCEEDED(overrides?: CallOverrides): Promise<[string]>;

    UNBACKED_MINT_CAP_EXCEEDED(overrides?: CallOverrides): Promise<[string]>;

    UNDERLYING_BALANCE_ZERO(overrides?: CallOverrides): Promise<[string]>;

    UNDERLYING_CANNOT_BE_RESCUED(overrides?: CallOverrides): Promise<[string]>;

    UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<[string]>;

    USER_IN_ISOLATION_MODE(overrides?: CallOverrides): Promise<[string]>;

    VARIABLE_DEBT_SUPPLY_NOT_ZERO(overrides?: CallOverrides): Promise<[string]>;

    ZERO_ADDRESS_NOT_VALID(overrides?: CallOverrides): Promise<[string]>;
  };

  ACL_ADMIN_CANNOT_BE_ZERO(overrides?: CallOverrides): Promise<string>;

  ADDRESSES_PROVIDER_ALREADY_ADDED(overrides?: CallOverrides): Promise<string>;

  ADDRESSES_PROVIDER_NOT_REGISTERED(overrides?: CallOverrides): Promise<string>;

  AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE(
    overrides?: CallOverrides
  ): Promise<string>;

  ASSET_NOT_BORROWABLE_IN_ISOLATION(overrides?: CallOverrides): Promise<string>;

  ASSET_NOT_LISTED(overrides?: CallOverrides): Promise<string>;

  BORROWING_NOT_ENABLED(overrides?: CallOverrides): Promise<string>;

  BORROW_CAP_EXCEEDED(overrides?: CallOverrides): Promise<string>;

  BRIDGE_PROTOCOL_FEE_INVALID(overrides?: CallOverrides): Promise<string>;

  CALLER_MUST_BE_POOL(overrides?: CallOverrides): Promise<string>;

  CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN(
    overrides?: CallOverrides
  ): Promise<string>;

  CALLER_NOT_ATOKEN(overrides?: CallOverrides): Promise<string>;

  CALLER_NOT_BRIDGE(overrides?: CallOverrides): Promise<string>;

  CALLER_NOT_EMERGENCY_ADMIN(overrides?: CallOverrides): Promise<string>;

  CALLER_NOT_POOL_ADMIN(overrides?: CallOverrides): Promise<string>;

  CALLER_NOT_POOL_CONFIGURATOR(overrides?: CallOverrides): Promise<string>;

  CALLER_NOT_POOL_OR_EMERGENCY_ADMIN(
    overrides?: CallOverrides
  ): Promise<string>;

  CALLER_NOT_RISK_OR_POOL_ADMIN(overrides?: CallOverrides): Promise<string>;

  COLLATERAL_BALANCE_IS_ZERO(overrides?: CallOverrides): Promise<string>;

  COLLATERAL_CANNOT_BE_LIQUIDATED(overrides?: CallOverrides): Promise<string>;

  COLLATERAL_CANNOT_COVER_NEW_BORROW(
    overrides?: CallOverrides
  ): Promise<string>;

  COLLATERAL_SAME_AS_BORROWING_CURRENCY(
    overrides?: CallOverrides
  ): Promise<string>;

  DEBT_CEILING_EXCEEDED(overrides?: CallOverrides): Promise<string>;

  DEBT_CEILING_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

  EMODE_CATEGORY_RESERVED(overrides?: CallOverrides): Promise<string>;

  FLASHLOAN_DISABLED(overrides?: CallOverrides): Promise<string>;

  FLASHLOAN_PREMIUM_INVALID(overrides?: CallOverrides): Promise<string>;

  HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD(
    overrides?: CallOverrides
  ): Promise<string>;

  HEALTH_FACTOR_NOT_BELOW_THRESHOLD(overrides?: CallOverrides): Promise<string>;

  INCONSISTENT_EMODE_CATEGORY(overrides?: CallOverrides): Promise<string>;

  INCONSISTENT_FLASHLOAN_PARAMS(overrides?: CallOverrides): Promise<string>;

  INCONSISTENT_PARAMS_LENGTH(overrides?: CallOverrides): Promise<string>;

  INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET(
    overrides?: CallOverrides
  ): Promise<string>;

  INVALID_ADDRESSES_PROVIDER(overrides?: CallOverrides): Promise<string>;

  INVALID_ADDRESSES_PROVIDER_ID(overrides?: CallOverrides): Promise<string>;

  INVALID_AMOUNT(overrides?: CallOverrides): Promise<string>;

  INVALID_BORROW_CAP(overrides?: CallOverrides): Promise<string>;

  INVALID_BURN_AMOUNT(overrides?: CallOverrides): Promise<string>;

  INVALID_DEBT_CEILING(overrides?: CallOverrides): Promise<string>;

  INVALID_DECIMALS(overrides?: CallOverrides): Promise<string>;

  INVALID_EMODE_CATEGORY(overrides?: CallOverrides): Promise<string>;

  INVALID_EMODE_CATEGORY_ASSIGNMENT(overrides?: CallOverrides): Promise<string>;

  INVALID_EMODE_CATEGORY_PARAMS(overrides?: CallOverrides): Promise<string>;

  INVALID_EXPIRATION(overrides?: CallOverrides): Promise<string>;

  INVALID_FLASHLOAN_EXECUTOR_RETURN(overrides?: CallOverrides): Promise<string>;

  INVALID_INTEREST_RATE_MODE_SELECTED(
    overrides?: CallOverrides
  ): Promise<string>;

  INVALID_LIQUIDATION_PROTOCOL_FEE(overrides?: CallOverrides): Promise<string>;

  INVALID_LIQ_BONUS(overrides?: CallOverrides): Promise<string>;

  INVALID_LIQ_THRESHOLD(overrides?: CallOverrides): Promise<string>;

  INVALID_LTV(overrides?: CallOverrides): Promise<string>;

  INVALID_MINT_AMOUNT(overrides?: CallOverrides): Promise<string>;

  INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO(
    overrides?: CallOverrides
  ): Promise<string>;

  INVALID_OPTIMAL_USAGE_RATIO(overrides?: CallOverrides): Promise<string>;

  INVALID_RESERVE_FACTOR(overrides?: CallOverrides): Promise<string>;

  INVALID_RESERVE_INDEX(overrides?: CallOverrides): Promise<string>;

  INVALID_RESERVE_PARAMS(overrides?: CallOverrides): Promise<string>;

  INVALID_SIGNATURE(overrides?: CallOverrides): Promise<string>;

  INVALID_SUPPLY_CAP(overrides?: CallOverrides): Promise<string>;

  INVALID_UNBACKED_MINT_CAP(overrides?: CallOverrides): Promise<string>;

  LTV_VALIDATION_FAILED(overrides?: CallOverrides): Promise<string>;

  NOT_CONTRACT(overrides?: CallOverrides): Promise<string>;

  NOT_ENOUGH_AVAILABLE_USER_BALANCE(overrides?: CallOverrides): Promise<string>;

  NO_DEBT_OF_SELECTED_TYPE(overrides?: CallOverrides): Promise<string>;

  NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF(
    overrides?: CallOverrides
  ): Promise<string>;

  NO_MORE_RESERVES_ALLOWED(overrides?: CallOverrides): Promise<string>;

  NO_OUTSTANDING_STABLE_DEBT(overrides?: CallOverrides): Promise<string>;

  NO_OUTSTANDING_VARIABLE_DEBT(overrides?: CallOverrides): Promise<string>;

  OPERATION_NOT_SUPPORTED(overrides?: CallOverrides): Promise<string>;

  POOL_ADDRESSES_DO_NOT_MATCH(overrides?: CallOverrides): Promise<string>;

  PRICE_ORACLE_SENTINEL_CHECK_FAILED(
    overrides?: CallOverrides
  ): Promise<string>;

  RESERVE_ALREADY_ADDED(overrides?: CallOverrides): Promise<string>;

  RESERVE_ALREADY_INITIALIZED(overrides?: CallOverrides): Promise<string>;

  RESERVE_DEBT_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

  RESERVE_FROZEN(overrides?: CallOverrides): Promise<string>;

  RESERVE_INACTIVE(overrides?: CallOverrides): Promise<string>;

  RESERVE_LIQUIDITY_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

  RESERVE_PAUSED(overrides?: CallOverrides): Promise<string>;

  SILOED_BORROWING_VIOLATION(overrides?: CallOverrides): Promise<string>;

  SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER(
    overrides?: CallOverrides
  ): Promise<string>;

  STABLE_BORROWING_ENABLED(overrides?: CallOverrides): Promise<string>;

  STABLE_BORROWING_NOT_ENABLED(overrides?: CallOverrides): Promise<string>;

  STABLE_DEBT_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

  SUPPLY_CAP_EXCEEDED(overrides?: CallOverrides): Promise<string>;

  UNBACKED_MINT_CAP_EXCEEDED(overrides?: CallOverrides): Promise<string>;

  UNDERLYING_BALANCE_ZERO(overrides?: CallOverrides): Promise<string>;

  UNDERLYING_CANNOT_BE_RESCUED(overrides?: CallOverrides): Promise<string>;

  UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO(
    overrides?: CallOverrides
  ): Promise<string>;

  USER_IN_ISOLATION_MODE(overrides?: CallOverrides): Promise<string>;

  VARIABLE_DEBT_SUPPLY_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

  ZERO_ADDRESS_NOT_VALID(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    ACL_ADMIN_CANNOT_BE_ZERO(overrides?: CallOverrides): Promise<string>;

    ADDRESSES_PROVIDER_ALREADY_ADDED(
      overrides?: CallOverrides
    ): Promise<string>;

    ADDRESSES_PROVIDER_NOT_REGISTERED(
      overrides?: CallOverrides
    ): Promise<string>;

    AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE(
      overrides?: CallOverrides
    ): Promise<string>;

    ASSET_NOT_BORROWABLE_IN_ISOLATION(
      overrides?: CallOverrides
    ): Promise<string>;

    ASSET_NOT_LISTED(overrides?: CallOverrides): Promise<string>;

    BORROWING_NOT_ENABLED(overrides?: CallOverrides): Promise<string>;

    BORROW_CAP_EXCEEDED(overrides?: CallOverrides): Promise<string>;

    BRIDGE_PROTOCOL_FEE_INVALID(overrides?: CallOverrides): Promise<string>;

    CALLER_MUST_BE_POOL(overrides?: CallOverrides): Promise<string>;

    CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN(
      overrides?: CallOverrides
    ): Promise<string>;

    CALLER_NOT_ATOKEN(overrides?: CallOverrides): Promise<string>;

    CALLER_NOT_BRIDGE(overrides?: CallOverrides): Promise<string>;

    CALLER_NOT_EMERGENCY_ADMIN(overrides?: CallOverrides): Promise<string>;

    CALLER_NOT_POOL_ADMIN(overrides?: CallOverrides): Promise<string>;

    CALLER_NOT_POOL_CONFIGURATOR(overrides?: CallOverrides): Promise<string>;

    CALLER_NOT_POOL_OR_EMERGENCY_ADMIN(
      overrides?: CallOverrides
    ): Promise<string>;

    CALLER_NOT_RISK_OR_POOL_ADMIN(overrides?: CallOverrides): Promise<string>;

    COLLATERAL_BALANCE_IS_ZERO(overrides?: CallOverrides): Promise<string>;

    COLLATERAL_CANNOT_BE_LIQUIDATED(overrides?: CallOverrides): Promise<string>;

    COLLATERAL_CANNOT_COVER_NEW_BORROW(
      overrides?: CallOverrides
    ): Promise<string>;

    COLLATERAL_SAME_AS_BORROWING_CURRENCY(
      overrides?: CallOverrides
    ): Promise<string>;

    DEBT_CEILING_EXCEEDED(overrides?: CallOverrides): Promise<string>;

    DEBT_CEILING_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

    EMODE_CATEGORY_RESERVED(overrides?: CallOverrides): Promise<string>;

    FLASHLOAN_DISABLED(overrides?: CallOverrides): Promise<string>;

    FLASHLOAN_PREMIUM_INVALID(overrides?: CallOverrides): Promise<string>;

    HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<string>;

    HEALTH_FACTOR_NOT_BELOW_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<string>;

    INCONSISTENT_EMODE_CATEGORY(overrides?: CallOverrides): Promise<string>;

    INCONSISTENT_FLASHLOAN_PARAMS(overrides?: CallOverrides): Promise<string>;

    INCONSISTENT_PARAMS_LENGTH(overrides?: CallOverrides): Promise<string>;

    INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET(
      overrides?: CallOverrides
    ): Promise<string>;

    INVALID_ADDRESSES_PROVIDER(overrides?: CallOverrides): Promise<string>;

    INVALID_ADDRESSES_PROVIDER_ID(overrides?: CallOverrides): Promise<string>;

    INVALID_AMOUNT(overrides?: CallOverrides): Promise<string>;

    INVALID_BORROW_CAP(overrides?: CallOverrides): Promise<string>;

    INVALID_BURN_AMOUNT(overrides?: CallOverrides): Promise<string>;

    INVALID_DEBT_CEILING(overrides?: CallOverrides): Promise<string>;

    INVALID_DECIMALS(overrides?: CallOverrides): Promise<string>;

    INVALID_EMODE_CATEGORY(overrides?: CallOverrides): Promise<string>;

    INVALID_EMODE_CATEGORY_ASSIGNMENT(
      overrides?: CallOverrides
    ): Promise<string>;

    INVALID_EMODE_CATEGORY_PARAMS(overrides?: CallOverrides): Promise<string>;

    INVALID_EXPIRATION(overrides?: CallOverrides): Promise<string>;

    INVALID_FLASHLOAN_EXECUTOR_RETURN(
      overrides?: CallOverrides
    ): Promise<string>;

    INVALID_INTEREST_RATE_MODE_SELECTED(
      overrides?: CallOverrides
    ): Promise<string>;

    INVALID_LIQUIDATION_PROTOCOL_FEE(
      overrides?: CallOverrides
    ): Promise<string>;

    INVALID_LIQ_BONUS(overrides?: CallOverrides): Promise<string>;

    INVALID_LIQ_THRESHOLD(overrides?: CallOverrides): Promise<string>;

    INVALID_LTV(overrides?: CallOverrides): Promise<string>;

    INVALID_MINT_AMOUNT(overrides?: CallOverrides): Promise<string>;

    INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO(
      overrides?: CallOverrides
    ): Promise<string>;

    INVALID_OPTIMAL_USAGE_RATIO(overrides?: CallOverrides): Promise<string>;

    INVALID_RESERVE_FACTOR(overrides?: CallOverrides): Promise<string>;

    INVALID_RESERVE_INDEX(overrides?: CallOverrides): Promise<string>;

    INVALID_RESERVE_PARAMS(overrides?: CallOverrides): Promise<string>;

    INVALID_SIGNATURE(overrides?: CallOverrides): Promise<string>;

    INVALID_SUPPLY_CAP(overrides?: CallOverrides): Promise<string>;

    INVALID_UNBACKED_MINT_CAP(overrides?: CallOverrides): Promise<string>;

    LTV_VALIDATION_FAILED(overrides?: CallOverrides): Promise<string>;

    NOT_CONTRACT(overrides?: CallOverrides): Promise<string>;

    NOT_ENOUGH_AVAILABLE_USER_BALANCE(
      overrides?: CallOverrides
    ): Promise<string>;

    NO_DEBT_OF_SELECTED_TYPE(overrides?: CallOverrides): Promise<string>;

    NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF(
      overrides?: CallOverrides
    ): Promise<string>;

    NO_MORE_RESERVES_ALLOWED(overrides?: CallOverrides): Promise<string>;

    NO_OUTSTANDING_STABLE_DEBT(overrides?: CallOverrides): Promise<string>;

    NO_OUTSTANDING_VARIABLE_DEBT(overrides?: CallOverrides): Promise<string>;

    OPERATION_NOT_SUPPORTED(overrides?: CallOverrides): Promise<string>;

    POOL_ADDRESSES_DO_NOT_MATCH(overrides?: CallOverrides): Promise<string>;

    PRICE_ORACLE_SENTINEL_CHECK_FAILED(
      overrides?: CallOverrides
    ): Promise<string>;

    RESERVE_ALREADY_ADDED(overrides?: CallOverrides): Promise<string>;

    RESERVE_ALREADY_INITIALIZED(overrides?: CallOverrides): Promise<string>;

    RESERVE_DEBT_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

    RESERVE_FROZEN(overrides?: CallOverrides): Promise<string>;

    RESERVE_INACTIVE(overrides?: CallOverrides): Promise<string>;

    RESERVE_LIQUIDITY_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

    RESERVE_PAUSED(overrides?: CallOverrides): Promise<string>;

    SILOED_BORROWING_VIOLATION(overrides?: CallOverrides): Promise<string>;

    SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER(
      overrides?: CallOverrides
    ): Promise<string>;

    STABLE_BORROWING_ENABLED(overrides?: CallOverrides): Promise<string>;

    STABLE_BORROWING_NOT_ENABLED(overrides?: CallOverrides): Promise<string>;

    STABLE_DEBT_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

    SUPPLY_CAP_EXCEEDED(overrides?: CallOverrides): Promise<string>;

    UNBACKED_MINT_CAP_EXCEEDED(overrides?: CallOverrides): Promise<string>;

    UNDERLYING_BALANCE_ZERO(overrides?: CallOverrides): Promise<string>;

    UNDERLYING_CANNOT_BE_RESCUED(overrides?: CallOverrides): Promise<string>;

    UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<string>;

    USER_IN_ISOLATION_MODE(overrides?: CallOverrides): Promise<string>;

    VARIABLE_DEBT_SUPPLY_NOT_ZERO(overrides?: CallOverrides): Promise<string>;

    ZERO_ADDRESS_NOT_VALID(overrides?: CallOverrides): Promise<string>;
  };

  filters: {};

  estimateGas: {
    ACL_ADMIN_CANNOT_BE_ZERO(overrides?: CallOverrides): Promise<BigNumber>;

    ADDRESSES_PROVIDER_ALREADY_ADDED(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    ADDRESSES_PROVIDER_NOT_REGISTERED(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    ASSET_NOT_BORROWABLE_IN_ISOLATION(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    ASSET_NOT_LISTED(overrides?: CallOverrides): Promise<BigNumber>;

    BORROWING_NOT_ENABLED(overrides?: CallOverrides): Promise<BigNumber>;

    BORROW_CAP_EXCEEDED(overrides?: CallOverrides): Promise<BigNumber>;

    BRIDGE_PROTOCOL_FEE_INVALID(overrides?: CallOverrides): Promise<BigNumber>;

    CALLER_MUST_BE_POOL(overrides?: CallOverrides): Promise<BigNumber>;

    CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    CALLER_NOT_ATOKEN(overrides?: CallOverrides): Promise<BigNumber>;

    CALLER_NOT_BRIDGE(overrides?: CallOverrides): Promise<BigNumber>;

    CALLER_NOT_EMERGENCY_ADMIN(overrides?: CallOverrides): Promise<BigNumber>;

    CALLER_NOT_POOL_ADMIN(overrides?: CallOverrides): Promise<BigNumber>;

    CALLER_NOT_POOL_CONFIGURATOR(overrides?: CallOverrides): Promise<BigNumber>;

    CALLER_NOT_POOL_OR_EMERGENCY_ADMIN(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    CALLER_NOT_RISK_OR_POOL_ADMIN(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    COLLATERAL_BALANCE_IS_ZERO(overrides?: CallOverrides): Promise<BigNumber>;

    COLLATERAL_CANNOT_BE_LIQUIDATED(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    COLLATERAL_CANNOT_COVER_NEW_BORROW(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    COLLATERAL_SAME_AS_BORROWING_CURRENCY(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    DEBT_CEILING_EXCEEDED(overrides?: CallOverrides): Promise<BigNumber>;

    DEBT_CEILING_NOT_ZERO(overrides?: CallOverrides): Promise<BigNumber>;

    EMODE_CATEGORY_RESERVED(overrides?: CallOverrides): Promise<BigNumber>;

    FLASHLOAN_DISABLED(overrides?: CallOverrides): Promise<BigNumber>;

    FLASHLOAN_PREMIUM_INVALID(overrides?: CallOverrides): Promise<BigNumber>;

    HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    HEALTH_FACTOR_NOT_BELOW_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INCONSISTENT_EMODE_CATEGORY(overrides?: CallOverrides): Promise<BigNumber>;

    INCONSISTENT_FLASHLOAN_PARAMS(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INCONSISTENT_PARAMS_LENGTH(overrides?: CallOverrides): Promise<BigNumber>;

    INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INVALID_ADDRESSES_PROVIDER(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_ADDRESSES_PROVIDER_ID(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INVALID_AMOUNT(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_BORROW_CAP(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_BURN_AMOUNT(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_DEBT_CEILING(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_DECIMALS(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_EMODE_CATEGORY(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_EMODE_CATEGORY_ASSIGNMENT(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INVALID_EMODE_CATEGORY_PARAMS(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INVALID_EXPIRATION(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_FLASHLOAN_EXECUTOR_RETURN(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INVALID_INTEREST_RATE_MODE_SELECTED(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INVALID_LIQUIDATION_PROTOCOL_FEE(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INVALID_LIQ_BONUS(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_LIQ_THRESHOLD(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_LTV(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_MINT_AMOUNT(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    INVALID_OPTIMAL_USAGE_RATIO(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_RESERVE_FACTOR(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_RESERVE_INDEX(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_RESERVE_PARAMS(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_SIGNATURE(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_SUPPLY_CAP(overrides?: CallOverrides): Promise<BigNumber>;

    INVALID_UNBACKED_MINT_CAP(overrides?: CallOverrides): Promise<BigNumber>;

    LTV_VALIDATION_FAILED(overrides?: CallOverrides): Promise<BigNumber>;

    NOT_CONTRACT(overrides?: CallOverrides): Promise<BigNumber>;

    NOT_ENOUGH_AVAILABLE_USER_BALANCE(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    NO_DEBT_OF_SELECTED_TYPE(overrides?: CallOverrides): Promise<BigNumber>;

    NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    NO_MORE_RESERVES_ALLOWED(overrides?: CallOverrides): Promise<BigNumber>;

    NO_OUTSTANDING_STABLE_DEBT(overrides?: CallOverrides): Promise<BigNumber>;

    NO_OUTSTANDING_VARIABLE_DEBT(overrides?: CallOverrides): Promise<BigNumber>;

    OPERATION_NOT_SUPPORTED(overrides?: CallOverrides): Promise<BigNumber>;

    POOL_ADDRESSES_DO_NOT_MATCH(overrides?: CallOverrides): Promise<BigNumber>;

    PRICE_ORACLE_SENTINEL_CHECK_FAILED(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    RESERVE_ALREADY_ADDED(overrides?: CallOverrides): Promise<BigNumber>;

    RESERVE_ALREADY_INITIALIZED(overrides?: CallOverrides): Promise<BigNumber>;

    RESERVE_DEBT_NOT_ZERO(overrides?: CallOverrides): Promise<BigNumber>;

    RESERVE_FROZEN(overrides?: CallOverrides): Promise<BigNumber>;

    RESERVE_INACTIVE(overrides?: CallOverrides): Promise<BigNumber>;

    RESERVE_LIQUIDITY_NOT_ZERO(overrides?: CallOverrides): Promise<BigNumber>;

    RESERVE_PAUSED(overrides?: CallOverrides): Promise<BigNumber>;

    SILOED_BORROWING_VIOLATION(overrides?: CallOverrides): Promise<BigNumber>;

    SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    STABLE_BORROWING_ENABLED(overrides?: CallOverrides): Promise<BigNumber>;

    STABLE_BORROWING_NOT_ENABLED(overrides?: CallOverrides): Promise<BigNumber>;

    STABLE_DEBT_NOT_ZERO(overrides?: CallOverrides): Promise<BigNumber>;

    SUPPLY_CAP_EXCEEDED(overrides?: CallOverrides): Promise<BigNumber>;

    UNBACKED_MINT_CAP_EXCEEDED(overrides?: CallOverrides): Promise<BigNumber>;

    UNDERLYING_BALANCE_ZERO(overrides?: CallOverrides): Promise<BigNumber>;

    UNDERLYING_CANNOT_BE_RESCUED(overrides?: CallOverrides): Promise<BigNumber>;

    UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    USER_IN_ISOLATION_MODE(overrides?: CallOverrides): Promise<BigNumber>;

    VARIABLE_DEBT_SUPPLY_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    ZERO_ADDRESS_NOT_VALID(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    ACL_ADMIN_CANNOT_BE_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    ADDRESSES_PROVIDER_ALREADY_ADDED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    ADDRESSES_PROVIDER_NOT_REGISTERED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    ASSET_NOT_BORROWABLE_IN_ISOLATION(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    ASSET_NOT_LISTED(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    BORROWING_NOT_ENABLED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    BORROW_CAP_EXCEEDED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    BRIDGE_PROTOCOL_FEE_INVALID(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    CALLER_MUST_BE_POOL(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    CALLER_NOT_ATOKEN(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    CALLER_NOT_BRIDGE(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    CALLER_NOT_EMERGENCY_ADMIN(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    CALLER_NOT_POOL_ADMIN(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    CALLER_NOT_POOL_CONFIGURATOR(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    CALLER_NOT_POOL_OR_EMERGENCY_ADMIN(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    CALLER_NOT_RISK_OR_POOL_ADMIN(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    COLLATERAL_BALANCE_IS_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    COLLATERAL_CANNOT_BE_LIQUIDATED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    COLLATERAL_CANNOT_COVER_NEW_BORROW(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    COLLATERAL_SAME_AS_BORROWING_CURRENCY(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    DEBT_CEILING_EXCEEDED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    DEBT_CEILING_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    EMODE_CATEGORY_RESERVED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    FLASHLOAN_DISABLED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    FLASHLOAN_PREMIUM_INVALID(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    HEALTH_FACTOR_NOT_BELOW_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INCONSISTENT_EMODE_CATEGORY(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INCONSISTENT_FLASHLOAN_PARAMS(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INCONSISTENT_PARAMS_LENGTH(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_ADDRESSES_PROVIDER(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_ADDRESSES_PROVIDER_ID(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_AMOUNT(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    INVALID_BORROW_CAP(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_BURN_AMOUNT(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_DEBT_CEILING(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_DECIMALS(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    INVALID_EMODE_CATEGORY(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_EMODE_CATEGORY_ASSIGNMENT(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_EMODE_CATEGORY_PARAMS(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_EXPIRATION(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_FLASHLOAN_EXECUTOR_RETURN(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_INTEREST_RATE_MODE_SELECTED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_LIQUIDATION_PROTOCOL_FEE(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_LIQ_BONUS(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    INVALID_LIQ_THRESHOLD(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_LTV(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    INVALID_MINT_AMOUNT(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_OPTIMAL_USAGE_RATIO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_RESERVE_FACTOR(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_RESERVE_INDEX(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_RESERVE_PARAMS(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_SIGNATURE(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    INVALID_SUPPLY_CAP(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    INVALID_UNBACKED_MINT_CAP(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    LTV_VALIDATION_FAILED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    NOT_CONTRACT(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    NOT_ENOUGH_AVAILABLE_USER_BALANCE(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    NO_DEBT_OF_SELECTED_TYPE(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    NO_MORE_RESERVES_ALLOWED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    NO_OUTSTANDING_STABLE_DEBT(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    NO_OUTSTANDING_VARIABLE_DEBT(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    OPERATION_NOT_SUPPORTED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    POOL_ADDRESSES_DO_NOT_MATCH(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    PRICE_ORACLE_SENTINEL_CHECK_FAILED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    RESERVE_ALREADY_ADDED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    RESERVE_ALREADY_INITIALIZED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    RESERVE_DEBT_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    RESERVE_FROZEN(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    RESERVE_INACTIVE(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    RESERVE_LIQUIDITY_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    RESERVE_PAUSED(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    SILOED_BORROWING_VIOLATION(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    STABLE_BORROWING_ENABLED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    STABLE_BORROWING_NOT_ENABLED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    STABLE_DEBT_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    SUPPLY_CAP_EXCEEDED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    UNBACKED_MINT_CAP_EXCEEDED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    UNDERLYING_BALANCE_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    UNDERLYING_CANNOT_BE_RESCUED(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    UNDERLYING_CLAIMABLE_RIGHTS_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    USER_IN_ISOLATION_MODE(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    VARIABLE_DEBT_SUPPLY_NOT_ZERO(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    ZERO_ADDRESS_NOT_VALID(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
