// @aave/core-v3/types/factories/flashloan/interfaces/IFlashLoanSimpleReceiver__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {Contract, Signer, utils} from "ethers";
import type {Provider} from "@ethersproject/providers";
import type {
  IFlashLoanSimpleReceiver,
  IFlashLoanSimpleReceiverInterface,
} from "../../../flashloan/interfaces/IFlashLoanSimpleReceiver";

const _abi = [
  {
    inputs: [],
    name: "ADDRESSES_PROVIDER",
    outputs: [
      {
        internalType: "contract IPoolAddressesProvider",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "POOL",
    outputs: [
      {
        internalType: "contract IPool",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "asset",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "premium",
        type: "uint256",
      },
      {
        internalType: "address",
        name: "initiator",
        type: "address",
      },
      {
        internalType: "bytes",
        name: "params",
        type: "bytes",
      },
    ],
    name: "executeOperation",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class IFlashLoanSimpleReceiver__factory {
  static readonly abi = _abi;
  static createInterface(): IFlashLoanSimpleReceiverInterface {
    return new utils.Interface(_abi) as IFlashLoanSimpleReceiverInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IFlashLoanSimpleReceiver {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as IFlashLoanSimpleReceiver;
  }
}
